content/src/content/jcr_root/apps/core/wcm/components/commons/datalayer/clientlibs/js/datalayer.js

```diff
@@ -0,0 +1,170 @@
+            updateState(window.dataLayer.state, event);
+            return Array.prototype.push.apply(this, arguments);
+        };
+    }
+
+    function updateState(state, event) {
+        if (event && event.type === "updated") {
+            deepMerge(state, event.object);
+        }
+    }
+
+    function deepMerge(target, source) {
+        var tmpSource = {};
+        if (isObject(target) && isObject(source)) {
+            Object.keys(source).forEach(function(key) {
+                if (isObject(source[key])) {
+                    if (!target[key]) {
+                        tmpSource[key] = {};
+                        Object.assign(target, tmpSource);
+                    }
+                    deepMerge(target[key], source[key]);
+                } else {
+                    tmpSource[key] = source[key];
+                    Object.assign(target, tmpSource);
+                }
+            });
+        }
+    }
+
+    function isObject(item) {
+        return (item && typeof item === "object" && !Array.isArray(item));
+    }
+
+    function populateDataLayer() {
+        window.dataLayer.push({
+            "type": "updated",
+            "object": {
+                "component": {
+                    "carousel": {
+                        "carousel3": {
+                            "id": "/content/mysite/en/home/jcr:content/root/carousel3",
+                            "items": {}
+                        }
+                    }
+                }
+            }
+        });
+
+        window.dataLayer.push({
+            "type": "updated",
+            "object": {
+                "component": {
+                    "tab": {
+                        "tab2": {
+                            "id": "/content/mysite/en/home/jcr:content/root/tab2",
+                            "items": {}
+                        }
+                    }
+                }
+            }
+        });
+
+        window.dataLayer.push({
+            "type": "updated",
+            "object": {
+                "page": {
+                    "id": "/content/mysite/en/products/crossfit",
+                    "siteLanguage": "en-us",
+                    "siteCountry": "US",
+                    "pageType": "product detail",
+                    "pageName": "pdp - crossfit zoom",
+                    "pageCategory": "womens > shoes > athletic"
+                }
+            }
+        });
+
+        window.dataLayer.push({
+            "type": "listenerdefined",
+            "object": {
+                "listener": {
+                    "type": "clicked",
+                    "target": {},
+                    "handler": function(component) {
+                        console.log("clicked on component: " + component);
+                    }
+                }
+            }
+        });
+    }
+
+    function populateDataLayerAfterOverride() {
+        window.dataLayer.push({
+            "type": "updated",
+            "object": {
```

Comment:
I'm not keen on this "object" key, while not a reserved word in JavaScript I think we can find something better.

Reply:
how about "data" or "branch"? "details" sounds a bit weird to me.

Reply:
"data" works for me